src/envoy/http/authn/http_filter.cc

```diff
@@ -30,59 +57,211 @@ void AuthenticationFilter::onDestroy() {
   ENVOY_LOG(debug, "Called AuthenticationFilter : {}", __func__);
 }
 
-FilterHeadersStatus AuthenticationFilter::decodeHeaders(HeaderMap&, bool) {
+FilterHeadersStatus AuthenticationFilter::decodeHeaders(HeaderMap& headers,
+                                                        bool) {
   ENVOY_LOG(debug, "Called AuthenticationFilter : {}", __func__);
+  state_ = IstioAuthN::State::PROCESSING;
+  if (config_.peers_size() == 0) {
+    ENVOY_LOG(debug, "No method defined. Skip source authentication.");
+    onAuthenticatePeerDone(&headers, 0, nullptr, IstioAuthN::Status::SUCCESS);
+  } else {
+    authenticatePeer(
+        headers, config_.peers(0),
+        std::bind(&AuthenticationFilter::onAuthenticatePeerDone, this, &headers,
+                  0, std::placeholders::_1, std::placeholders::_2));
+  }
+  if (state_ == IstioAuthN::State::COMPLETE) {
+    return FilterHeadersStatus::Continue;
+  }
 
-  int peer_size = config_.peers_size();
-  ENVOY_LOG(debug, "AuthenticationFilter: {} config.peers_size()={}", __func__,
-            peer_size);
-  if (peer_size > 0) {
-    const ::istio::authentication::v1alpha1::Mechanism& m = config_.peers()[0];
-    if (m.has_mtls()) {
-      ENVOY_LOG(debug, "AuthenticationFilter: {} this connection requires mTLS",
-                __func__);
-      MtlsAuthentication mtls_authn(decoder_callbacks_->connection());
-      if (mtls_authn.IsMutualTLS() == false) {
-        // In prototype, only log the authentication policy violation.
-        ENVOY_LOG(error,
-                  "AuthenticationFilter: authn policy requires mTLS but the "
-                  "connection is not mTLS!");
+  stopped_ = true;
+  return FilterHeadersStatus::StopIteration;
+}
+
+void AuthenticationFilter::authenticatePeer(
+    HeaderMap& headers, const iaapi::PeerAuthenticationMethod& method,
+    const AuthenticateDoneCallback& done_callback) {
+  switch (method.params_case()) {
+    case iaapi::PeerAuthenticationMethod::ParamsCase::kMtls:
+      ValidateX509(headers, method.mtls(), done_callback);
+      break;
+    case iaapi::PeerAuthenticationMethod::ParamsCase::kJwt:
+      ValidateJwt(headers, method.jwt(), done_callback);
+      break;
+    case iaapi::PeerAuthenticationMethod::ParamsCase::kNone:
+      done_callback(nullptr, IstioAuthN::Status::SUCCESS);
+      break;
+    default:
+      ENVOY_LOG(error, "Unknown peer authentication param {}",
+                method.DebugString());
+  }
+}
+
+void AuthenticationFilter::onAuthenticatePeerDone(
+    HeaderMap* headers, int peer_method_index,
+    std::unique_ptr<IstioAuthN::AuthenticatePayload> payload,
+    const IstioAuthN::Status& status) {
```

Comment:
For Status &status, only its value is used here. It will be safer for the Status &status be passed by value, not passed by reference.

Reply:
Change this to bool anyway. We don't use any information from 'authenticator' call yet.